subprojects/core-api/src/main/java/org/gradle/api/artifacts/transform/ArtifactTransformOutputs.java

```diff
@@ -17,24 +17,27 @@
 package org.gradle.api.artifacts.transform;
 
 import org.gradle.api.Incubating;
-import org.gradle.api.reflect.InjectionPointQualifier;
+import org.gradle.internal.HasInternalProtocol;
 
 import java.io.File;
-import java.lang.annotation.Documented;
-import java.lang.annotation.ElementType;
-import java.lang.annotation.Retention;
-import java.lang.annotation.RetentionPolicy;
-import java.lang.annotation.Target;
 
 /**
- * Attached to a property that should receive the <em>workspace</em> for an artifact transform. This is the directory into which the transform should generate its output files.
+ * The outputs of the artifact transform.
  *
- * @since 5.2
+ * @since 5.3
  */
 @Incubating
-@Retention(RetentionPolicy.RUNTIME)
-@Target({ElementType.METHOD})
-@Documented
-@InjectionPointQualifier(supportedTypes = File.class)
-public @interface Workspace {
+@HasInternalProtocol
+public interface ArtifactTransformOutputs {
+    /**
+     * Registers an output file of the transform.
+     *
+     * <p>The order of calls to this method is retained in the result of the transform.</p>
+     */
+    void registerOutput(File outputFile);
```

Comment:
What do we think about adding something like:

```java
File registerOutput(String relativePath);
```

My guess would be that people will be creating files like this:

```java
File output = new File(outputs.getWorkspace(), "output.txt");
outputs.registerOutput(output);
// ... work with the output
```

Which seems a bit redundant, and we could do it like this instead:

```java
File output = outputs.registerOutput("output.txt");
// ... work with the output
```

We could even remove `getWorkspace()`—why would we need to expose its location as a thing?

Reply:
I did add the `getWorkspace()` since quite a few transforms use the workspace directly as an output. We could fix the API by adding
```
File registerOutput(String relativePath)
File registerOutput()
```

where the second method would return the workspace directory.

I'll try it out.

Reply:
Just after adding the new methods I realized we also need some way to return files in the _primary input_ as _outputs_. So we still would need a method `registerOutput(File outputFile)` to cover this use case. Am I missing something?

Reply:
We discussed this today, and I agree we need to keep that method.